packages/ksf-registration/src/KSF/Registration/Component.purs

```diff
@@ -61,50 +67,75 @@ data RegistrationInputField
-  , onChange :: Maybe String -> Effect Unit
-  , value :: Maybe String
-  }
-
-type InputType = String
-
-data Validation = Valid | Invalid
-derive instance eqValidation :: Eq Validation
-
-data EmailValidation = InUse | Validation Validation
+type ValidationFn = Maybe String -> Maybe ValidationError
 
-type Pattern = String
+data ValidationError
+  = Invalid String
+  | InvalidEmpty String
+  | InvalidPatternFailure String
+  | InvalidEmailInUse String
+  | InvalidNotInitialized -- Fictional state only to be set when the form is first rendered
+derive instance eqValidationError :: Eq ValidationError
 
 data Action
   = UpdateInput RegistrationInputField (Maybe String)
-  | PasswordMismatch Validation
-  | EmailInUse
-  | FormFieldInvalid String
+  | FormFieldValidation RegistrationInputField
+  | SetServerValidationError RegistrationInputField ValidationError
 
 registration :: Props -> JSX
 registration = make component { initialState, render, update }
 
 initialState :: State
 initialState =
-  { firstName: Nothing
-  , lastName: Nothing
-  , streetAddress: Nothing
-  , city: Nothing
-  , zip: Nothing
-  , country: Just "FI"
-  , phone: Nothing
-  , emailAddress: Nothing
-  , password: Nothing
-  , confirmPassword: Nothing
-  , inputValidations:
-     { password: Valid
-     , passwordConfirm: Valid
-     , emailAddress: Validation Valid
-     }
+  { firstName: emptyField { validationFns = [validateEmptyField FirstName firstNameMissing] }
+  , lastName: emptyField { validationFns = [validateEmptyField LastName lastNameMissing] }
+  , streetAddress: emptyField { validationFns = [validateEmptyField StreetAddress streetAddressMissing] }
+  , city: emptyField { validationFns = [validateEmptyField City cityMissing] }
+  , zip: emptyField { validationFns = [validateEmptyField Zip zipMissing, validateInputWithRegex Zip zipRegexPattern zipRegexFailure] }
+  , country: emptyField { value = Just "FI", validationError = Nothing }
+  , phone: emptyField { validationFns = [validateEmptyField Phone phoneMissing, validateInputWithRegex Phone phoneRegexPattern phoneRegexFailure] }
+  , emailAddress: emptyField { validationFns = [validateEmptyField EmailAddress emailMissing, validateInputWithRegex EmailAddress emailRegex emailRegexFailure] }
+  , password: emptyField { validationFns = [validatePassword] }
+  , confirmPassword: emptyField { validationFns = [validateEmptyField ConfirmPassword passwordMissing] }
   }
+  where
+    emptyField :: InputField
+    emptyField = { value: Nothing, validationError: Just $ InvalidNotInitialized, validationFns: [] }
+
+    -- Allows word characters (a-z, A-Z, 0-9, _), dashes and whitespaces
+    zipRegexPattern = "^[\\s|\\w|-]+$"
+    zipRegexFailure = "Postnummerfältet kan bara innehålla siffror och bokstäver."
+
+    -- Allows digits, "+" and whitespaces,
+    -- e.g. 040 1231234, +358 04 1231234
+    phoneRegexPattern = "^[\\d|\\+|\\s|-|\\(|\\)]+$"
+    phoneRegexFailure = "Telefonnummer kan bara innehålla nummer."
```

Comment:
Small detail: should be "Telefonnummer kan bara bestå av siffror."
Or maybe: "Telefonnummer kan bara bestå av siffror, mellanslag och +-tecken." since we allow those as well.

Reply:
Changed to "Telefonnummer kan bara bestå av siffror, mellanslag och +-tecken." as it's indeed more accurate.